cachi2/core/package_managers/pip.py

```diff
@@ -1648,136 +1814,97 @@ def _validate_provided_hashes(requirements: list[PipRequirement], require_hashes
+    Process wheel artifacts.
+
+    A note on nomenclature (to address a common misconception)
+    It really is not "source" vs. "wheel" (i.e. "wheel" is somehow inherently not "source")
+    It is very much "sdist" vs. "wheel".
+    To a first approximation, sdists and wheels are BOTH "source".
+    - sdists are usually "raw" repo dumps, including docs, build scripts, etc.
+    - wheels are *only* Python source (but see below) - so, there's the irony: wheels are _usually_
+      more "source" than sdists
+
+    The confusion probably arises because wheels allow and encourage binary artifacts; in reality,
+    it's not that common.
+    This is why I use the term "sdist" instead of "source" (which is horribly overloaded anyway),
+    and in fact PyPI's own "kind" attributes are "sdist" and "wheel".
+
+    :param requirement: which pip package to process
+    :param str pip_deps_dir:
+    :param bool allow_binary: process wheels?
+    :return: a single DistributionPackageInfo, or a list of DPI
+    :rtype: DistributionPackageInfo
+    """
+    allowed_distros = ["sdist", "wheel"] if allow_binary else ["sdist"]
+    best_sdist: list[DistributionPackageInfo] = []
+    client = pypi_simple.PyPISimple()
     name = requirement.package
     version = requirement.version_specs[0][1]
     normalized_version = canonicalize_version(version)
+    sdists: list[DistributionPackageInfo] = []
+    user_checksums = set(map(_to_checksum_info, requirement.hashes))
+    wheels: list[DistributionPackageInfo] = []
 
-    client = pypi_simple.PyPISimple()
     try:
         timeout = get_config().requests_timeout
         project_page = client.get_project_page(name, timeout)
-        packages = project_page.packages
+        packages: list[pypi_simple.DistributionPackage] = project_page.packages
     except (requests.RequestException, pypi_simple.NoSuchProjectError) as e:
         raise FetchError(f"PyPI query failed: {e}")
 
-    allowed_distros = ["sdist", "wheel"] if allow_binary else ["sdist"]
-    filtered_packages = filter(
-        lambda x: x.version is not None
-        and canonicalize_version(x.version) == normalized_version
-        and x.package_type is not None
-        and x.package_type in allowed_distros,
-        packages,
-    )
-
-    sdists: list[DistributionPackageInfo] = []
-    wheels: list[DistributionPackageInfo] = []
+    def _is_valid(pkg: pypi_simple.DistributionPackage) -> bool:
+        return (
+            pkg.version is not None
+            and canonicalize_version(pkg.version) == normalized_version
+            and pkg.package_type is not None
+            and pkg.package_type in allowed_distros
+        )
 
-    user_checksums = set(map(_to_checksum_info, requirement.hashes))
+    valid_packages = (package for package in packages if _is_valid(package))
 
-    for package in filtered_packages:
-        pypi_checksums = {
+    for package in valid_packages:
+        pypi_checksums: set[ChecksumInfo] = {
```

Comment:
maybe `valid_packages` variable is not needed here as well, I can imagine something like this instead:

```
for package in packages:
    if not _is_valid(...):
        continue
```